src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelModuleResolutionFunction.java

```diff
@@ -108,48 +117,64 @@ public SkyValue compute(SkyKey skyKey, Environment env)
     SkyframeLookupResult repoSpecResults = env.getValuesAndExceptions(repoSpecKeys);
     ImmutableMap.Builder<ModuleKey, RepoSpec> remoteRepoSpecs = ImmutableMap.builder();
     for (RepoSpecKey repoSpecKey : repoSpecKeys) {
-      RepoSpec repoSpec = (RepoSpec) repoSpecResults.get(repoSpecKey);
-      if (repoSpec == null) {
+      RepoSpecValue repoSpecValue = (RepoSpecValue) repoSpecResults.get(repoSpecKey);
+      if (repoSpecValue == null) {
         return null;
       }
-      remoteRepoSpecs.put(repoSpecKey.getModuleKey(), repoSpec);
+      remoteRepoSpecs.put(repoSpecKey.getModuleKey(), repoSpecValue.repoSpec());
+      repoSpecValue
+          .registryFileHashes()
+          .forEach(
+              (url, checksum) ->
+                  // We only ever download the same file from the same registry once, so
+                  // mismatching checksums can only occur with crafted setups (e.g., two
+                  // --registry flags that only differ by a terminating slash and with contents
+                  // that change during a single build). We crash in this case.
+                  registryFileHashes.merge(
+                      url, checksum, RegistryFileDownloadEvent::failOnDifferentChecksums));
     }
 
     ImmutableMap<ModuleKey, Module> finalDepGraph;
     try (SilentCloseable c =
         Profiler.instance().profile(ProfilerTask.BZLMOD, "compute final dep graph")) {
       finalDepGraph =
           computeFinalDepGraph(
-              state.selectionResult.getResolvedDepGraph(),
+              state.discoverAndSelectResult.selectionResult.getResolvedDepGraph(),
               root.getOverrides(),
               remoteRepoSpecs.buildOrThrow());
     }
 
     return BazelModuleResolutionValue.create(
-        finalDepGraph, state.selectionResult.getUnprunedDepGraph());
+        finalDepGraph,
+        state.discoverAndSelectResult.selectionResult.getUnprunedDepGraph(),
+        ImmutableMap.copyOf(registryFileHashes));
   }
 
+  private record Result(
+      Selection.Result selectionResult,
```

Comment:
`BazelModuleResolutionFunction.Result` is `Selection.Result + ChecksumMap`, but `Selection.Result` itself is `DepGraph + ChecksumMap`. Could we just unify these two into `DepGraphWithChecksums` or something?

Reply:
`Discovery.Result` is `DepGraph + ChecksumMap`, `Selection.Result` is `DepGraph + DepGraph` and `Result` is `Selection.Result + ChecksumMap`. I also dislike this setup, but I'm not sure how to resolve this without squashing everything into a `DepGraph + Optional<DepGraph> + (possibly empty) ChecksumMap` and losing type safety in the process.